File: JCON/KotlinMigration/src/main/java/nl/blue4it/basic/controller/PaymentController.java
package nl.blue4it.basic.controller;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import nl.blue4it.basic.controller.dto.request.Payment;
import nl.blue4it.basic.service.PaymentService;
import nl.blue4it.basic.service.exception.PaymentRejectedException;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;

import javax.validation.Valid;
import java.util.List;

@RestController
@Slf4j
@RequiredArgsConstructor
@Validated
public class PaymentController {
    private final PaymentService paymentService;
    @PostMapping
    public ResponseEntity<String> processPayment(@Valid @RequestBody List<Payment> payment) throws PaymentRejectedException {
        log.info("Processing payments");

        try {
            paymentService.processPayment(payment);
            log.info("Payment processed");
            return ResponseEntity.ok("Created payment");
        } catch (PaymentRejectedException ex) {
            log.info("Got an error when processing payment");
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
        }
    }
}

File: JCON/KotlinMigration/src/main/java/nl/blue4it/basic/controller/PaymentController.kt
package nl.blue4it.basic.controller

import mu.KotlinLogging
import nl.blue4it.basic.controller.dto.request.Payment
import nl.blue4it.basic.service.PaymentService


class PaymentController constructor(
    private val paymentService: PaymentService
) {
    private val log = KotlinLogging.logger {}

    suspend fun processPayment(payments: List<Payment>): String {
        return kotlin.runCatching {
            log.info("Processing payments")
            paymentService.processPayment(payments)
        }.onFailure {
            log.info("Got an error when processing payment")
            // metrics
        }.onSuccess {
            log.info("Payment processed")
            // metrics
        }.fold(
            {  "Created payment" },
            {  "Something failed" },
        )
    }
}